src/databricks/labs/ucx/migration/azure_credentials.py

```diff
@@ -0,0 +1,249 @@
+        #     ],
+        #   system assigned managed identity should have
+        #     "alternativeNames": [
+        #         "isExplicit=False",
+        #         "/subscriptions/3f2e4d32-8e8d-46d6-82bc-5bb8d962328b/resourcegroups/pepsi_rg/providers/Microsoft.Databricks/accessConnectors/ag-conn-access-connector"
+        #     ],
+        # Service Principal should have "servicePrincipalType": "Application"
+        #
+        # update if_mi field of the sp in sp_list
+        # TODO: find out if azure databricks access connector with system assigned managed identity
+        # TODO: has role assignment on storage, what ID will be returned by providers/Microsoft.Authorization/roleAssignments
+        return
+
+
+    def _list_storage_credentials(self):
+        # list existed storage credentials
+        # for SP storage credentials, capture its application_id
+        # for MI storage credentials:
+        #   if managed_identity_id is not empty capture it, this will be the id of the user assigned managed identity, like:
+        #       /subscriptions/3f2e4d32-8e8d-46d6-82bc-5bb8d962328b/resourcegroups/ziyuanqin-uc-test/providers/Microsoft.ManagedIdentity/userAssignedIdentities/ziyuan-user-assigned-mi
+        #   else capture the access_connector_id, this is the resource id of the access connector, like:
+        #       /subscriptions/3f2e4d32-8e8d-46d6-82bc-5bb8d962328b/resourceGroups/pepsi_rg/providers/Microsoft.Databricks/accessConnectors/ag-conn-access-connector
+        # TODO: UC only has access_connector_id (resource id) for access connector with system assigned managed identity
+        # TODO: need to find the mapping between the access_connector_id and principalId from providers/Microsoft.Authorization/roleAssignments
+        return {}
+
+
+    def _check_sp_in_storage_credentials(self, sp_list, sc_set):
+        # if sp is already used, take it off from the sp_list
+        # TODO: PrincipalID in
+        return list()
+
+
+    def _fetch_client_secret(self, sp_list: list[ServicePrincipalMigrationInfo]):
+        # check AzureServicePrincipalInfo from AzureServicePrincipalCrawler, if AzureServicePrincipalInfo
+        # has secret_scope and secret_key not empty, fetch the client_secret and put it to the
+        # client_secret field
+
+        # fetch client_secrets of crawled service principal, if any
+        azure_sp_info_with_client_secret = {}
+        azure_sp_infos = self._azure_sp_crawler.snapshot()
+
+        for azure_sp_info in azure_sp_infos:
+            if azure_sp_info.secret_scope and azure_sp_info.secret_key:
+                try:
+                    secret_response = self._ws.secrets.get_secret(azure_sp_info.secret_scope, azure_sp_info.secret_key)
+
+                    # decode the binary string from GetSecretResponse to utf-8 string
+                    # TODO: handle different encoding if we have feedback from the customer
+                    binary_value_str = secret_response.value
```

Comment:
See ws.dbutils.secrets.get() method

Reply:
I have mistake here. I misread the bytes value as binary value, I just need to base64 decode the value, will fix it. 

But the comments for encoding still applies, customer may provide the bytes value instead of string value when create the secrets, and the provided bytes value may came from a different encoding than utf-8 and cause trouble when decode them. For example a bytes value like `T2zhLCBNdW5kbyE=` will not be correctly decode by `ws.dbutils.secrets.get()` as [it assume the charset is 'utf-8'](https://src.dev.databricks.com/databricks/runtime@master/-/blob/sql/core/src/main/scala/com/databricks/service/DBUtils.scala?L270:7&popover=pinned):
```scala
  def get(scope: String, key: String): String = {
    new String(getBytes(scope, key), Charsets.UTF_8)
  }
```
But we don't need to address it now, as I see the possibility of hitting this issue is extremely low.

Reply:
updated the code to base64 decode the secret value